log/devlog/devlog.go

```diff
@@ -0,0 +1,127 @@
+//
+// So while this package shouldn't be used for production, it can be enabled
+// for your development use by setting TIDEPOOL_USE_DEVLOG=1 as an environment
+// variable. It will also be enabled if POD_NAMESPACE=default (this is set via
+// the development repo when using it's kind k8s cluster).
+
+package devlog
+
+import (
+	"fmt"
+	"io"
+	"log"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/tidepool-org/platform/errors"
+	plog "github.com/tidepool-org/platform/log"
+)
+
+// New provides a plain-text log.Logger suitable for development or testing.
+func New(writer io.Writer, levelRanks plog.LevelRanks, level plog.Level) (plog.Logger, error) {
+	if writer == nil {
+		return nil, errors.New("writer is missing")
+	}
+	logLogger := log.New(writer, "", 0)
+	return plog.NewLogger(&serializer{logLogger}, levelRanks, level)
+}
+
+func NewWithDefaults(writer io.Writer) (plog.Logger, error) {
+	return New(writer, plog.DefaultLevelRanks(), plog.DebugLevel)
+}
+
+type serializer struct {
+	*log.Logger
+}
+
+func (s *serializer) Serialize(fields plog.Fields) error {
+	var pairs = make([]string, 0, len(fields))
+	var msg, msgTime, msgLevel, msgCaller string
+	var showCaller = false
+
+	msgLevel = abbreviateLevel("")
+	msgTime = time.Now().Format(time.Stamp)
+
+	for key, value := range fields {
+
+		switch key {
+		case "caller":
+			msgCaller = formatCaller(value)
+			continue
+		case "level":
+			level := sOrPlusV(value)
+			msgLevel = abbreviateLevel(level)
+			if level == "error" {
+				showCaller = true
+			}
+			continue
+		case "message":
+			msg = sOrPlusV(value)
+			continue
+		case "process":
+			// process isn't useful outside of sumo
+			continue
+		case "time":
+			// time is terser and constant width for console logging
+			tv, err := time.Parse(time.RFC3339Nano, sOrPlusV(value))
+			if err != nil {
+				msgTime = sOrPlusV(value)
+				continue
+			}
+			msgTime = tv.Format(time.Stamp)
+			continue
+		}
+		pairs = append(pairs, key+"="+sOrPlusV(value))
+	}
+	if showCaller {
+		pairs = append(pairs, msgCaller)
+	}
+	sort.Strings(pairs)
+	rest := ""
+	if len(pairs) > 0 {
+		rest = ": " + strings.Join(pairs, " ")
+	}
+	s.Logger.Printf(msgTime + " " + msgLevel + " " + msg + rest)
+	return nil
+}
+
+func formatCaller(value any) string {
+	cc, ok := value.(*errors.Caller)
+	if !ok {
+		return "caller=" + sOrPlusV(value)
+	}
+	return fmt.Sprintf("caller=%s:%d", cc.File, cc.Line)
+}
+
+func sOrPlusV(thing any) (s string) {
+	//defer func() { s = strings.TrimSpace(s) }()
+	if stringer, ok := thing.(fmt.Stringer); ok {
+		return stringer.String()
+	} else if s, ok := thing.(string); ok {
+		return s
+	}
+	return fmt.Sprintf("%+v", thing)
+}
+
+var abbr map[string]string = map[string]string{
+	"debug": "DD",
+	"error": "EE",
+	"info":  "II",
+	"warn":  "WW",
+}
+
+func abbreviateLevel(fullLevel string) string {
+	if a, ok := abbr[fullLevel]; ok {
+		return a
+	}
+	if fullLevel == "" {
+		return "??"
+	}
+	return strings.ToUpper(fullLevel[0:2])
```

Comment:
I'd just use "??" if not one of the known levels. (There is the edge case where the fullLevel only has one character.)

Reply:
Sure, doesn't matter much anyway, and not sure how it'd come about anyway.

Good catch on the edge case.